### PR TITLE
PUBDEV-6410: Isolation Forest - early stopping on mean anomaly score

### PR DESCRIPTION
**Pull request by michalkurka**
This PR adds support for early stopping on mean `anomaly_score` in Isolation Forest. In order to do that, I had to refactor ScoreKeeper quite a bit: adding both @wendycwong and @maurever to the review.

### CODE REVIEW

File: h2o-py/h2o/estimators/deeplearning.py [hunk 1]
```diff
@@ -1000,18 +1000,19 @@ def stopping_rounds(self, stopping_rounds):
...
-        and custom_increasing can only be used in GBM and DRF with the Python client.
+        Metric to use for early stopping (AUTO: logloss for classification, deviance for regression and anonomaly_score
+        for Isolation Forest). Note that custom and custom_increasing can only be used in GBM and DRF with the Python
+        client.
 
         One of: ``"auto"``, ``"deviance"``, ``"logloss"``, ``"mse"``, ``"rmse"``, ``"mae"``, ``"rmsle"``, ``"auc"``,
-        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"custom"``, ``"custom_increasing"``
-        (default: ``"auto"``).
+        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"anomaly_score"``, ``"custom"``,
+        ``"custom_increasing"``  (default: ``"auto"``).
```
Comment (maurever) [inline h2o-py/h2o/estimators/deeplearning.py]: Can be `anomaly_score` be used for other algos? If not, does it make sense to have `anomaly_score` in other algos as the possibility if a user can't use it?

File: h2o-core/src/main/java/hex/ScoreKeeper.java [hunk 2]
```diff
@@ -111,30 +111,133 @@ else if (m instanceof ModelMetricsMultinomial) {
...
+    MAE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    RMSLE(ConvergenceStrategy.LESS_IS_BETTER, true),
+    AUC(ConvergenceStrategy.MORE_IS_BETTER, false),
+    lift_top_group(ConvergenceStrategy.MORE_IS_BETTER, false),
+    misclassification(ConvergenceStrategy.LESS_IS_BETTER, true),
+    mean_per_class_error(ConvergenceStrategy.LESS_IS_BETTER, true),
+    custom(ConvergenceStrategy.LESS_IS_BETTER, false),
+    custom_increasing(ConvergenceStrategy.MORE_IS_BETTER, false),
+    anomaly_score(ConvergenceStrategy.NON_DIRECTIONAL, false);
+
```
Comment (wendycwong) [inline h2o-core/src/main/java/hex/ScoreKeeper.java]: Why is anomaly_score NON_DIRECTIONAL?  How do you determine if you are doing better or not if you do not know the direction it should go then?

File: h2o-py/h2o/estimators/deepwater.py [hunk 3]
```diff
@@ -637,18 +637,19 @@ def stopping_rounds(self, stopping_rounds):
...
+        Metric to use for early stopping (AUTO: logloss for classification, deviance for regression and anonomaly_score
+        for Isolation Forest). Note that custom and custom_increasing can only be used in GBM and DRF with the Python
+        client.
 
         One of: ``"auto"``, ``"deviance"``, ``"logloss"``, ``"mse"``, ``"rmse"``, ``"mae"``, ``"rmsle"``, ``"auc"``,
-        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"custom"``, ``"custom_increasing"``
-        (default: ``"auto"``).
+        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"anomaly_score"``, ``"custom"``,
+        ``"custom_increasing"``  (default: ``"auto"``).
         """
```
Comment (wendycwong) [inline h2o-py/h2o/estimators/deepwater.py]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-py/h2o/estimators/random_forest.py [hunk 4]
```diff
@@ -481,18 +481,19 @@ def stopping_rounds(self, stopping_rounds):
...
-        and custom_increasing can only be used in GBM and DRF with the Python client.
+        Metric to use for early stopping (AUTO: logloss for classification, deviance for regression and anonomaly_score
+        for Isolation Forest). Note that custom and custom_increasing can only be used in GBM and DRF with the Python
+        client.
 
         One of: ``"auto"``, ``"deviance"``, ``"logloss"``, ``"mse"``, ``"rmse"``, ``"mae"``, ``"rmsle"``, ``"auc"``,
-        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"custom"``, ``"custom_increasing"``
-        (default: ``"auto"``).
+        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"anomaly_score"``, ``"custom"``,
+        ``"custom_increasing"``  (default: ``"auto"``).
```
Comment (wendycwong) [inline h2o-py/h2o/estimators/random_forest.py]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-py/h2o/estimators/deeplearning.py (hunk 1 above)
Comment (michalkurka) [reply h2o-py/h2o/estimators/deeplearning.py]: It cannot but this is a current H2O limitation. This list is a union of everything available in H2O (eg. deep learning doesn't support custom metrics either).

Good point though.

File: h2o-py/h2o/estimators/xgboost.py [hunk 5]
```diff
@@ -282,18 +282,19 @@ def stopping_rounds(self, stopping_rounds):
...
-        and custom_increasing can only be used in GBM and DRF with the Python client.
+        Metric to use for early stopping (AUTO: logloss for classification, deviance for regression and anonomaly_score
+        for Isolation Forest). Note that custom and custom_increasing can only be used in GBM and DRF with the Python
+        client.
 
         One of: ``"auto"``, ``"deviance"``, ``"logloss"``, ``"mse"``, ``"rmse"``, ``"mae"``, ``"rmsle"``, ``"auc"``,
-        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"custom"``, ``"custom_increasing"``
-        (default: ``"auto"``).
+        ``"lift_top_group"``, ``"misclassification"``, ``"mean_per_class_error"``, ``"anomaly_score"``, ``"custom"``,
+        ``"custom_increasing"``  (default: ``"auto"``).
```
Comment (wendycwong) [inline h2o-py/h2o/estimators/xgboost.py]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-r/h2o-package/R/deeplearning.R [hunk 6]
```diff
@@ -208,7 +209,7 @@ h2o.deeplearning <- function(x, y, training_frame,
                              classification_stop = 0,
                              regression_stop = 1e-06,
                              stopping_rounds = 5,
-                             stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "custom", "custom_increasing"),
+                             stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "anomaly_score", "custom", "custom_increasing"),
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/deeplearning.R]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-r/h2o-package/R/deepwater.R [hunk 7]
```diff
@@ -141,7 +142,7 @@ h2o.deepwater <- function(x, y, training_frame,
                           classification_stop = 0,
                           regression_stop = 0,
                           stopping_rounds = 5,
-                          stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "custom", "custom_increasing"),
+                          stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "anomaly_score", "custom", "custom_increasing"),
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/deepwater.R]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-r/h2o-package/R/gbm.R [hunk 8]
```diff
@@ -141,7 +142,7 @@ h2o.gbm <- function(x, y, training_frame,
                     nbins_cats = 1024,
                     r2_stopping = 1.797693135e+308,
                     stopping_rounds = 0,
-                    stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "custom", "custom_increasing"),
+                    stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "anomaly_score", "custom", "custom_increasing"),
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/gbm.R]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-r/h2o-package/R/randomforest.R [hunk 9]
```diff
@@ -117,7 +118,7 @@ h2o.randomForest <- function(x, y, training_frame,
                              nbins_cats = 1024,
                              r2_stopping = 1.797693135e+308,
                              stopping_rounds = 0,
-                             stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "custom", "custom_increasing"),
+                             stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "anomaly_score", "custom", "custom_increasing"),
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/randomforest.R]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-r/h2o-package/R/xgboost.R [hunk 10]
```diff
@@ -106,7 +107,7 @@ h2o.xgboost <- function(x, y, training_frame,
                         offset_column = NULL,
                         weights_column = NULL,
                         stopping_rounds = 0,
-                        stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "custom", "custom_increasing"),
+                        stopping_metric = c("AUTO", "deviance", "logloss", "MSE", "RMSE", "MAE", "RMSLE", "AUC", "lift_top_group", "misclassification", "mean_per_class_error", "anomaly_score", "custom", "custom_increasing"),
```
Comment (wendycwong) [inline h2o-r/h2o-package/R/xgboost.R]: Same point as Veronica has pointed out.  Please fix the gen_python.py gen_R.py to get rid of anomaly_score for all other algos except isolation forest.

File: h2o-core/src/main/java/hex/ScoreKeeper.java [hunk 11]
```diff
@@ -111,30 +111,133 @@ else if (m instanceof ModelMetricsMultinomial) {
       _custom_metric =  m._custom_metric.value;
   }
 
-  public enum StoppingMetric { AUTO, deviance, logloss, MSE, RMSE,MAE,RMSLE, AUC, lift_top_group, misclassification, mean_per_class_error, custom, custom_increasing} //, r2}
-  static boolean moreIsBetter(StoppingMetric criterion) {
-    return (criterion == StoppingMetric.AUC || criterion == StoppingMetric.lift_top_group || criterion == StoppingMetric.custom_increasing); // || criterion == StoppingMetric.r2);
+  public interface IStoppingMetric {
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ScoreKeeper.java]: No rule, we don't do that anywhere else but there are also not many interfaces in H2O. I didn't want to rename the StoppingMetric enum (that would create tons of issue for users) so I was left with not that many choices for the interface name.

File: h2o-py/h2o/estimators/deepwater.py (hunk 3 above)
Comment (michalkurka) [reply h2o-py/h2o/estimators/deepwater.py]: This is not easily possible and we don't do that anywhere else. IMHO we shouldn't do that.

The way we address this is by documentation. Is that okay @wendycwong?

File: h2o-core/src/main/java/hex/ScoreKeeper.java (hunk 2 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ScoreKeeper.java]: Isolation Forest is unsupervised, there is no way to measure if it is doing better or worse. The early stopping is a based on the mean anomaly score stabilizing - that is it stops changing in either direction. It doesn't go "a lot" up or "a lot" down. That is how we tell the algo to stop early.

**Comment by michalkurka** [comment]
@maurever @wendycwong thanks for the review - I added the exclude for anomaly_score, it should show up only in IsolationForest

**Pull request merged**